From 7b70190094095f3ea2c4f05ef02639615fa582f8 Mon Sep 17 00:00:00 2001
From: cjoseph <cjoseph@meraki.com>
Date: Thu, 14 May 2015 14:41:53 -0700
Subject: [PATCH]     openwrt/syslogd logserver compatibility patch + startup
 script modifications

    This code adds a patch for busybox-1.20.2's system logging functionality so
    that log messages bound for a remote host contain the hostname. This allows
    a remote logging server to disambiguate between messages that come in via
    port forwarding.

    /etc/init.d/S10boot was also modified to look for the generic_syslog_servers
    configuration option and modify the syslogd to send remote logging messages
    to those servers

    Reviewed-by: pallas
---
 sysklogd/syslogd.c | 127 ++++++++++++++++++++++++---------------------
 1 file changed, 69 insertions(+), 58 deletions(-)

diff --git a/sysklogd/syslogd.c b/sysklogd/syslogd.c
index ab50f4a..6478915 100644
--- a/sysklogd/syslogd.c
+++ b/sysklogd/syslogd.c
@@ -125,7 +125,7 @@
 //usage:#define syslogd_trivial_usage
 //usage:       "[OPTIONS]"
 //usage:#define syslogd_full_usage "\n\n"
-//usage:       "System logging utility\n"
+//usage:       "System logging utility (patched)\n"
 //usage:	IF_NOT_FEATURE_SYSLOGD_CFG(
 //usage:       "(this version of syslogd ignores /etc/syslog.conf)\n"
 //usage:	)
@@ -288,6 +288,8 @@ struct globals {
 	 * host (64), fac.prio (20) to the message */
 	/* (growth by: 15 + 64 + 20 + delims = ~110) */
 	char printbuf[MAX_READ*2 + 128];
+	/* an additional buffer for adding hostname for remote syslog */
+	char netbuf[MAX_READ*2 + 128];
 };
 
 static const struct init_globals init_data = {
@@ -819,6 +821,24 @@ static void parse_fac_prio_20(int pri, char *res20)
 	snprintf(res20, 20, "<%d>", pri);
 }
 
+#if ENABLE_FEATURE_REMOTE_LOG
+static int try_to_resolve_remote(remoteHost_t *rh)
+{
+	if (!rh->remoteAddr) {
+		unsigned now = monotonic_sec();
+
+		/* Don't resolve name too often - DNS timeouts can be big */
+		if ((now - rh->last_dns_resolve) < DNS_WAIT_SEC)
+			return -1;
+		rh->last_dns_resolve = now;
+		rh->remoteAddr = host2sockaddr(rh->remoteHostname, 514);
+		if (!rh->remoteAddr)
+			return -1;
+	}
+	return xsocket(rh->remoteAddr->u.sa.sa_family, SOCK_DGRAM, 0);
+}
+#endif
+
 /* len parameter is used only for "is there a timestamp?" check.
  * NB: some callers cheat and supply len==0 when they know
  * that there is no timestamp, short-circuiting the test. */
@@ -870,8 +890,56 @@ static void timestamp_and_log(int pri, char *msg, int len)
 		char res[20];
 		parse_fac_prio_20(pri, res);
 		sprintf(G.printbuf, "%s %.64s %s %s\n", timestamp, G.hostname, res, msg);
+		sprintf(G.netbuf, "%s %.64s %s %s\n", timestamp, G.hostname, res, msg);
+	}
+
+	size_t net_sz;
+	net_sz = MAX_READ*2 + 128;
+
+	/* Drop trailing '\n' and NULs (typically there is one NUL) */
+	while (1) {
+		if (G.netbuf[net_sz-1] != '\0' && G.netbuf[net_sz-1] != '\n')
+			break;
+		net_sz--;
 	}
 
+#if ENABLE_FEATURE_REMOTE_LOG
+	llist_t *item;
+	/* Stock syslogd sends it '\n'-terminated
+	 * over network, mimic that */
+	G.netbuf[net_sz] = '\n';
+
+	/* We are not modifying log messages in any way before send */
+	/* Remote site cannot trust _us_ anyway and need to do validation again */
+	for (item = G.remoteHosts; item != NULL; item = item->link) {
+		remoteHost_t *rh = (remoteHost_t *)item->data;
+
+		if (rh->remoteFD == -1) {
+			rh->remoteFD = try_to_resolve_remote(rh);
+			if (rh->remoteFD == -1)
+				continue;
+		}
+
+		/* Send message to remote logger.
+		 * On some errors, close and set remoteFD to -1
+		 * so that DNS resolution is retried.
+		 */
+		if (sendto(rh->remoteFD, G.netbuf, net_sz+1,
+				MSG_DONTWAIT | MSG_NOSIGNAL,
+				&(rh->remoteAddr->u.sa), rh->remoteAddr->len) == -1
+		) {
+			switch (errno) {
+			case ECONNRESET:
+			case ENOTCONN: /* paranoia */
+			case EPIPE:
+				close(rh->remoteFD);
+				rh->remoteFD = -1;
+				free(rh->remoteAddr);
+				rh->remoteAddr = NULL;
+			}
+		}
+	}
+#endif
 	/* Log message locally (to file or shared mem) */
 #if ENABLE_FEATURE_SYSLOGD_CFG
 	{
@@ -985,30 +1053,9 @@ static NOINLINE int create_socket(void)
 	return sock_fd;
 }
 
-#if ENABLE_FEATURE_REMOTE_LOG
-static int try_to_resolve_remote(remoteHost_t *rh)
-{
-	if (!rh->remoteAddr) {
-		unsigned now = monotonic_sec();
-
-		/* Don't resolve name too often - DNS timeouts can be big */
-		if ((now - rh->last_dns_resolve) < DNS_WAIT_SEC)
-			return -1;
-		rh->last_dns_resolve = now;
-		rh->remoteAddr = host2sockaddr(rh->remoteHostname, 514);
-		if (!rh->remoteAddr)
-			return -1;
-	}
-	return xsocket(rh->remoteAddr->u.sa.sa_family, SOCK_DGRAM, 0);
-}
-#endif
-
 static void do_syslogd(void) NORETURN;
 static void do_syslogd(void)
 {
-#if ENABLE_FEATURE_REMOTE_LOG
-	llist_t *item;
-#endif
 #if ENABLE_FEATURE_SYSLOGD_DUP
 	int last_sz = -1;
 	char *last_buf;
@@ -1074,42 +1121,6 @@ static void do_syslogd(void)
 			if (memcmp(last_buf, recvbuf, sz) == 0)
 				continue;
 		last_sz = sz;
-#endif
-#if ENABLE_FEATURE_REMOTE_LOG
-		/* Stock syslogd sends it '\n'-terminated
-		 * over network, mimic that */
-		recvbuf[sz] = '\n';
-
-		/* We are not modifying log messages in any way before send */
-		/* Remote site cannot trust _us_ anyway and need to do validation again */
-		for (item = G.remoteHosts; item != NULL; item = item->link) {
-			remoteHost_t *rh = (remoteHost_t *)item->data;
-
-			if (rh->remoteFD == -1) {
-				rh->remoteFD = try_to_resolve_remote(rh);
-				if (rh->remoteFD == -1)
-					continue;
-			}
-
-			/* Send message to remote logger.
-			 * On some errors, close and set remoteFD to -1
-			 * so that DNS resolution is retried.
-			 */
-			if (sendto(rh->remoteFD, recvbuf, sz+1,
-					MSG_DONTWAIT | MSG_NOSIGNAL,
-					&(rh->remoteAddr->u.sa), rh->remoteAddr->len) == -1
-			) {
-				switch (errno) {
-				case ECONNRESET:
-				case ENOTCONN: /* paranoia */
-				case EPIPE:
-					close(rh->remoteFD);
-					rh->remoteFD = -1;
-					free(rh->remoteAddr);
-					rh->remoteAddr = NULL;
-				}
-			}
-		}
 #endif
 		if (!ENABLE_FEATURE_REMOTE_LOG || (option_mask32 & OPT_locallog)) {
 			recvbuf[sz] = '\0'; /* ensure it *is* NUL terminated */
-- 
2.32.0

